src/collision.cpp

```diff
@@ -466,96 +454,126 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
+		const aabb3f& cbox = nearest_info.box;
 
-			//movingbox except moved to the horizontal position it would be after step up
+		//movingbox except moved to the horizontal position it would be after step up
+		bool step_up = false;
+		if (nearest_collided != COLLISION_AXIS_Y) {
 			aabb3f stepbox = movingbox;
-			stepbox.MinEdge.X += speed_f->X * dtime;
-			stepbox.MinEdge.Z += speed_f->Z * dtime;
-			stepbox.MaxEdge.X += speed_f->X * dtime;
-			stepbox.MaxEdge.Z += speed_f->Z * dtime;
+			// Look slightly ahead  for checking the height when stepping
+			// to ensure we also check above the node we collided with
+			// otherwise, might allow glitches such as a stack of stairs
+			float extra_dtime = nearest_dtime + 0.1f * fabsf(dtime - nearest_dtime);
+			stepbox.MinEdge.X += aspeed_f.X * extra_dtime;
+			stepbox.MinEdge.Z += aspeed_f.Z * extra_dtime;
+			stepbox.MaxEdge.X += aspeed_f.X * extra_dtime;
+			stepbox.MaxEdge.Z += aspeed_f.Z * extra_dtime;
 			// Check for stairs.
-			bool step_up = (nearest_collided != COLLISION_AXIS_Y) && // must not be Y direction
-					(movingbox.MinEdge.Y < cbox.MaxEdge.Y) &&
-					(movingbox.MinEdge.Y + stepheight > cbox.MaxEdge.Y) &&
-					(!wouldCollideWithCeiling(cinfo, stepbox,
-							cbox.MaxEdge.Y - movingbox.MinEdge.Y,
-							d));
-
-			// Get bounce multiplier
-			float bounce = -(float)nearest_info.bouncy / 100.0f;
-
-			// Move to the point of collision and reduce dtime by nearest_dtime
-			if (nearest_dtime < 0) {
-				// Handle negative nearest_dtime
-				if (!step_up) {
-					if (nearest_collided == COLLISION_AXIS_X)
-						pos_f->X += speed_f->X * nearest_dtime;
-					if (nearest_collided == COLLISION_AXIS_Y)
-						pos_f->Y += speed_f->Y * nearest_dtime;
-					if (nearest_collided == COLLISION_AXIS_Z)
-						pos_f->Z += speed_f->Z * nearest_dtime;
-				}
-			} else {
-				*pos_f += truncate(*speed_f * nearest_dtime, 100.0f);
-				dtime -= nearest_dtime;
+			step_up = (movingbox.MinEdge.Y < cbox.MaxEdge.Y) &&
+				(movingbox.MinEdge.Y + stepheight > cbox.MaxEdge.Y) &&
+				(!wouldCollideWithCeiling(cinfo, stepbox,
+						cbox.MaxEdge.Y - movingbox.MinEdge.Y,
+						d));
+		}
+
+		// Get bounce multiplier
+		float bounce = -(float)nearest_info.bouncy / 100.0f;
+
+		// Move to the point of collision and reduce dtime by nearest_dtime
+		if (nearest_dtime < 0) {
+			// Handle negative nearest_dtime
+			// This largely means an "instant" collision, e.g., with the floor.
+			// We use aspeed and nearest_dtime to be consistent with above and resolve this collision
+			if (!step_up) {
+				if (nearest_collided == COLLISION_AXIS_X)
+					pos_f->X += aspeed_f.X * nearest_dtime;
+				if (nearest_collided == COLLISION_AXIS_Y)
+					pos_f->Y += aspeed_f.Y * nearest_dtime;
+				if (nearest_collided == COLLISION_AXIS_Z)
+					pos_f->Z += aspeed_f.Z * nearest_dtime;
 			}
+		} else if (nearest_dtime > 0) {
+			aspeed_f = *speed_f + accel_f * 0.5f * nearest_dtime;
```

Comment:
The 0.5 comes from adding the *average* acceleration over the interval `nearest_dtime`.
So it's fine as long as `nearest_dtime` is accurate (which it is not entirely, as discussed elsewhere).
As we only simulate up to `nearest_dtime` in this iteration, we have to update the average speed to match this segment only. We could iterate multiple times here to approximate this numerically (recheck the collision with the new aspeed until the change is tiny), but I believe this will be fine and is better fixed by improving the `axisAlignedCollision` function.
But when `nearest_dtime` is much smaller than `dtime`, then this version will be more accurate than not updating `aspeed_f` (because the initial `aspeed_f` then was not very accurate).